File: C#/328. Odd Even Linked List.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution
{
    public ListNode OddEvenList(ListNode head)
    {
        if (head == null || head.next == null)
            return head;
        ListNode odd = head, even = head.next, oTail = null, eHead = head.next;
        while (odd != null && even != null)
        {
            odd.next = even.next;
            even.next = odd.next == null ? null : odd.next.next;
            if (odd.next == null)
                oTail = odd;
            odd = odd.next;
            even = even.next;
        }
        if (oTail == null)
            odd.next = eHead;
        else oTail.next = eHead;
        return head;
    }
}

File: python/328. Odd Even Linked List.py
# -*- coding: utf-8 -*-

"""
@Author: xiezizhe
@Date: 25/6/2020 下午3:14
"""


class Solution:
    def oddEvenList(self, head: ListNode) -> ListNode:
        if not head:
            return None

        node, flag, even_head, even_tail = head.next, True, None, None
        odd_head = odd_tail = head
        while node:

            if flag:
                if not even_tail:
                    even_head = even_tail = node
                else:
                    even_tail.next = node
                    even_tail = node
            else:
                odd_tail.next = node
                odd_tail = node

            flag = not flag
            node = node.next

        if even_tail:
            even_tail.next = None
        odd_tail.next = even_head
        return head
